File: src/components/th2/Bai2.js
import React, { useState } from "react";
import ExcelImport from "../ExcelImport";

const TH2BAI2 = () => {
  const [data, setData] = useState([]);
  const [frequentItemsets, setFrequentItemsets] = useState([]);
  const [maximalItemsets, setMaximalItemsets] = useState([]);
  const [associationRules, setAssociationRules] = useState([]);
  const [maximalAssociationRules, setMaximalAssociationRules] = useState([]);

  const handleSubmit = () => {
    if (data.length === 0) {
      alert("Vui lòng tải dữ liệu trước khi tính toán.");
      return;
    }

    const transactions = data.map((row) =>
      Object.keys(row).filter((key) => key !== "TID" && typeof row[key] === "string" && row[key].trim() !== "")
    );

    if (transactions.length === 0) {
      alert("Dữ liệu không hợp lệ hoặc không có giao dịch nào.");
      return;
    }

    const oneHotEncoding = (transactions) => {
      const uniqueItems = [...new Set(transactions.flat())];
      console.log(uniqueItems);
      
      return transactions.map((transaction) =>
        uniqueItems.map((item) => transaction.includes(item))
      );
    };

    const encodedTransactions = oneHotEncoding(transactions);
    const uniqueItems = [...new Set(transactions.flat())];
    console.log(uniqueItems);
    

    const apriori = (encodedTransactions, minSupport) => {
      const frequentItemsets = [];
      uniqueItems.forEach((item, index) => {
        const support =
          encodedTransactions.filter((transaction) => transaction[index]).length / transactions.length;
        if (support >= minSupport) {
          frequentItemsets.push({
            itemsets: [item],
            support: support,
          });
        }
      });

      for (let i = 0; i < uniqueItems.length; i++) {
        for (let j = i + 1; j < uniqueItems.length; j++) {
          const support =
            encodedTransactions.filter(
              (transaction) => transaction[i] && transaction[j]
            ).length / transactions.length;

          if (support >= minSupport) {
            frequentItemsets.push({
              itemsets: [uniqueItems[i], uniqueItems[j]],
              support: support,
            });
          }
        }
      }

      return frequentItemsets;
    };

    const generateAssociationRules = (frequentItemsets, minConfidence) => {
      const rules = [];
      frequentItemsets.forEach((itemset) => {
        if (itemset.itemsets.length > 1) {
          const [item1, item2] = itemset.itemsets;
          const item1Support =
            encodedTransactions.filter(
              (transaction) => transaction[uniqueItems.indexOf(item1)]
            ).length / transactions.length;

          const rules1 = {
            antecedents: [item1],
            consequents: [item2],
            support: itemset.support,
            confidence: itemset.support / item1Support,
          };

          const rules2 = {
            antecedents: [item2],
            consequents: [item1],
            support: itemset.support,
            confidence:
              itemset.support /
              (encodedTransactions.filter(
                (transaction) => transaction[uniqueItems.indexOf(item2)]
              ).length / transactions.length),
          };

          if (rules1.confidence >= minConfidence) {
            rules.push(rules1);
          }
          if (rules2.confidence >= minConfidence) {
            rules.push(rules2);
          }
        }
      });

      return rules;
    };

    const isMaximal = (itemset, frequentItemsets) => {
      return !frequentItemsets.some(
        (other) =>
          itemset.itemsets.every((i) => other.itemsets.includes(i)) &&
          itemset.itemsets.length < other.itemsets.length
      );
    };

    const findMaximalItemsets = (frequentItemsets) => {
      return frequentItemsets.filter((itemset) => isMaximal(itemset, frequentItemsets));
    };

    const filterRulesFromMaximalItemsets = (rules, maximalItemsets) => {
      return rules.filter((rule) =>
        maximalItemsets.some((maximal) =>
          rule.antecedents.every((a) => maximal.itemsets.includes(a)) ||
          rule.consequents.every((c) => maximal.itemsets.includes(c))
        )
      );
    };

    const minSupport = 0.3;
    const minConfidence = 0.8;

    const frequent = apriori(encodedTransactions, minSupport);
    const maximal = findMaximalItemsets(frequent);
    const rules = generateAssociationRules(frequent, minConfidence);
    const maximalRules = filterRulesFromMaximalItemsets(rules, maximal);

    setFrequentItemsets(frequent);
    setMaximalItemsets(maximal);
    setAssociationRules(rules);
    setMaximalAssociationRules(maximalRules);
  };

  const renderTable = (data, columns, title) => (
    <div className="bg-white shadow-lg rounded-lg overflow-hidden mb-6 ">
      <div className="bg-blue-500 text-white px-4 py-3 font-bold text-lg">{title}</div>
      <div className="max-h-[400px] overflow-y-auto overflow-x-auto" 
        style={{ 
          scrollbarWidth: 'thin', 
          scrollbarColor: 'rgb(59 130 246) transparent' 
        }}>
        <table className="w-full">
          <thead className="bg-gray-100 border-b">
            <tr>
              {columns.map((col, index) => (
                <th
                  key={index}
                  className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                >
                  {col.charAt(0).toUpperCase() + col.slice(1)}
                </th>
              ))}
            </tr>
          </thead>
          <tbody className="divide-y divide-gray-200">
            {data.map((item, rowIndex) => (
              <tr key={rowIndex} className="hover:bg-gray-50 transition-colors duration-200">
                {columns.map((col, colIndex) => (
                  <td key={colIndex} className="px-6 py-4 whitespace-nowrap text-sm text-gray-900">
                    {col === "itemsets" || col === "antecedents" || col === "consequents"
                      ? `${item[col].join(", ")}`
                      : typeof item[col] === "number"
                      ? item[col].toFixed(4)
                      : item[col]}
                  </td>
                ))}
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </div>
  );

  return (
    <div className="min-h-screen p-8 relative">
      <div className="container mx-auto ">
        <h1 className="text-2xl font-extrabold text-center text-blue-600 mb-8 uppercase tracking-wide shadow-xs p-2">
        Phân Tích Tập Phổ Biến 2
        </h1>

        {!data.length && <ExcelImport data={data} setData={setData} />}
        {data.length > 0 && (
          <>
           { renderTable(
                        data,
                        Object.keys(data[0]),
                        "Dữ Liệu Nhập"
                    )}
            <div className="flex justify-center mb-6">
              <button
                onClick={handleSubmit}
                className="bg-blue-500 hover:bg-blue-400 text-white font-semibold py-3 px-6 rounded-full shadow-md transition-all duration-300 ease-in-out transform hover:-translate-y-1"
              >
                Thực hiện tính toán
              </button>
            </div>
            <div className="grid grid-cols-1 gap-6">
              {frequentItemsets.length > 0 && renderTable(frequentItemsets, ["itemsets", "support"], "Tập Phổ Biến")}
              {maximalItemsets.length > 0 && renderTable(maximalItemsets, ["itemsets", "support"], "Tập Phổ Biến Tối Đại")}
              {associationRules.length > 0 && renderTable(
                associationRules,
                ["antecedents", "consequents", "support", "confidence"],
                "Các Luật Kết Hợp"
              )}
              {maximalAssociationRules.length > 0 && renderTable(
                maximalAssociationRules,
                ["antecedents", "consequents", "support", "confidence"],
                "Các Luật Kết Hợp Từ Tập Phổ Biến Tối Đại"
              )}
            </div>
          </>
        )}
      </div>
    </div>
  );
};

export default TH2BAI2;

File: src/components/th2/bai2.py
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

# Bước 1: Dữ liệu giao dịch từ bảng bạn đã cung cấp
transactions = [
    ['COMPA', 'THƯỚC', 'TẬP TRẮNG'],               # T1
    ['KÉO', 'THƯỚC', 'TẬP TRẮNG', 'BÚT BI'],        # T2
    ['COMPA', 'THƯỚC', 'TẬP TRẮNG', 'BÚT BI'],      # T3
    ['KÉO', 'COMPA', 'TẬP TRẮNG', 'BÚT BI'],        # T4
    ['THƯỚC'],                                      # T5
    ['BÚT BI'],                                     # T6
    ['TẬP TRẮNG'],                                  # T7
    ['TẨY'],                                        # T8
    ['BÚT MÀU', 'TẨY'],                             # T9
    ['BÚT MÀU']                                     # T10
]

# Bước 2: One-hot encoding dữ liệu
te = TransactionEncoder()
te_ary = te.fit(transactions).transform(transactions)
df = pd.DataFrame(te_ary, columns=te.columns_)

# Bước 3: Áp dụng giải thuật Apriori với min_sup=0.3
frequent_itemsets = apriori(df, min_support=0.3, use_colnames=True)

# Bước 4: Tính toán các luật kết hợp với min_conf=0.8
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.8)

# Bước 5: Tìm tập phổ biến tối đại
def is_maximal(itemset, frequent_itemsets):
    """Kiểm tra xem một tập hợp có phải là tập phổ biến tối đại không"""
    for other_itemset in frequent_itemsets['itemsets']:
        if set(itemset).issubset(other_itemset) and set(itemset) != set(other_itemset):
            return False
    return True

maximal_itemsets = frequent_itemsets[frequent_itemsets['itemsets'].apply(lambda x: is_maximal(x, frequent_itemsets))]

# Bước 6: Lọc các luật mà các tập hợp tiên đề và hậu quả thuộc tập phổ biến tối đại
def is_from_maximal(rule, maximal_itemsets):
    """Kiểm tra xem tập antecedents và consequents có phải thuộc tập phổ biến tối đại"""
    antecedents = set(rule['antecedents'])
    consequents = set(rule['consequents'])
    
    for maximal in maximal_itemsets['itemsets']:
        if antecedents.issubset(maximal) or consequents.issubset(maximal):
            return True
    return False

# Lọc các luật kết hợp sao cho các antecedents hoặc consequents thuộc tập phổ biến tối đại
maximal_rules_filtered = rules[rules.apply(lambda rule: is_from_maximal(rule, maximal_itemsets), axis=1)]

# Hiển thị kết quả
print("Tập phổ biến:")
print(frequent_itemsets)

print("\nTập phổ biến tối đại:")
print(maximal_itemsets)

# Hiển thị các luật kết hợp từ tập phổ biến (chỉ hiển thị antecedents, consequents, support, confidence)
print("\nCác luật kết hợp từ tập phổ biến (chỉ hiển thị antecedents, consequents, support, confidence):")
rules_filtered = rules[['antecedents', 'consequents', 'support', 'confidence']]
print(rules_filtered)

# Hiển thị các luật kết hợp từ tập phổ biến tối đại với độ tin cậy >= 0.8 và bỏ cột conviction
print("\nCác luật kết hợp từ tập phổ biến tối đại với độ tin cậy >= 0.8 (chỉ hiển thị antecedents, consequents, support, confidence):")
maximal_rules_filtered_no_conviction = maximal_rules_filtered[['antecedents', 'consequents', 'support', 'confidence']]
print(maximal_rules_filtered_no_conviction)
